### PR TITLE
Base Transaction Processor

### PR DESCRIPTION
**Pull request by adamgering**
Adds a base transaction processor
Adds message handler registration
Changes the sawtooth client and state client
to be static classes.
Creates a singleton for the RBAC library

### CODE REVIEW

File: rbac/common/base/base_processor.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            "No handler registered for {}".format(
+                get_message_type_name(payload.message_type)
+            )
+        )
+    message_handlers[payload.message_type].apply(
+        header=header, payload=payload, state=state
+    )
+
+
+# pylint: disable=useless-object-inheritance
```
Comment (kthblmfld) [inline rbac/common/base/base_processor.py]: This Pylint catch might be onto something. Why is object being passed in here?

File: rbac/common/sawtooth/__init__.py [hunk 2]
```diff
@@ -12,3 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 # -----------------------------------------------------------------------------
+
+from rbac.common.sawtooth.client_sync import ClientSync
+
+# pylint: disable=invalid-name
+client = ClientSync()
+
+__all__ = ["client"]
```
Comment (kthblmfld) [inline rbac/common/sawtooth/__init__.py]: Why is this necessary? Are we doing a '*' import on client somewhere?

File: rbac/common/sawtooth/rbac_payload.py [hunk 3]
```diff
@@ -30,7 +41,6 @@ def make_payload(message, message_type, inputs, outputs):
     )
 
 
-# pylint: disable=no-member
 def unmake_payload(payload):
     """Turn a payload back into a message given it's message type"""
```
Comment (kthblmfld) [inline rbac/common/sawtooth/rbac_payload.py]: Nitpick: Grammar error in 'its'

File: rbac/processor/event_handler.py [hunk 4]
```diff
@@ -119,8 +125,19 @@ def namespaces(self):
...
+            payload = RBACPayload()
+            payload.ParseFromString(transaction.payload)
+
+            if self._processor.has_message_handler(message_type=payload.message_type):
+                return self._processor.handle_message(
+                    header=transaction.header, payload=payload, state=state
+                )
+        except ValueError as err:
+            raise InvalidTransaction(err)
+        except Exception as err:  # pylint: disable=broad-except
```
Comment (kthblmfld) [inline rbac/processor/event_handler.py]: Does this block require catching the generic Exception type? If not, we can allow it to break naturally on an unexpected error if/when that time comes and respond with specific exception handling.

File: rbac/common/base/base_message.py [hunk 5]
```diff
@@ -18,18 +18,19 @@
...
 
+# pylint: disable=too-many-public-methods
 class BaseMessage:
     def __init__(self):
-        """Objects and methods shared across message libraries"""
-        self.client = ClientSync()
-        self.state = StateClient()
+        """Base class for all messages"""
+        self._message_type_name = batcher.get_message_type_name(self.message_type)
+        processor.register_message_handler(self)
```
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/common/base/base_message.py]: The transaction processor executes the messages; thus it needs to know what they are.

File: rbac/common/base/base_message.py [hunk 6]
```diff
@@ -40,27 +41,37 @@ def getattr(self, item, attribute):
...
+        raise NotImplementedError("Class must implement this property")
+
+    @property
+    def message_type_name(self):
+        return self._message_type_name
 
     @property
     def message_proto(self):
-        raise NotImplementedError("Class must implement this method")
+        """The protobuf used to serialize this message type"""
```
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/common/base/base_message.py]: ?

File: rbac/common/base/base_message.py [hunk 7]
```diff
@@ -137,7 +148,7 @@ def send(self, signer_keypair, payload, object_id=None, target_id=None):
         )
         got = None
 
-        status = self.client.send_batches_get_status(batch_list=batch_list)
+        status = client.send_batches_get_status(batch_list=batch_list)
```
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/common/base/base_message.py]: create method needs to send the message to the transaction processor

File: rbac/common/sawtooth/__init__.py (hunk 2 above)
Comment (adamgering) [reply rbac/common/sawtooth/__init__.py]: marks client as a variable for purpose of export, not an unused variable

File: rbac/common/sawtooth/rbac_payload.py (hunk 3 above)
Comment (adamgering) [reply rbac/common/sawtooth/rbac_payload.py]: will fix nitpick in later PR

File: rbac/processor/event_handler.py (hunk 4 above)
Comment (adamgering) [reply rbac/processor/event_handler.py]: broad exception is required, transaction processor can't fail because of a message failure

File: rbac/common/base/base_message.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/common/base/base_message.py]: Previously the transaction processor discovered the message handlers. However, you explicitly disliked that pattern, thus the above is an alternative solution.

File: rbac/common/base/base_message.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/common/base/base_message.py]: This PR only added docstrings and message_type_name.

File: rbac/common/base/base_message.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/common/base/base_message.py]: This PR only added docstrings and a dependency reference in this class.

File: rbac/common/sawtooth/batcher.py [hunk 8]
```diff
@@ -23,9 +23,14 @@
 from rbac.app.config import BATCHER_KEY_PAIR
 from rbac.common.sawtooth import rbac_payload
 
+
 LOGGER = logging.getLogger(__name__)
 
 
+def get_message_type_name(message_type):
+    return rbac_payload.get_message_type_name(message_type=message_type)
```
Comment (nadiabahrami) [inline rbac/common/sawtooth/batcher.py]: Docstring please

File: rbac/common/sawtooth/state_client.py [hunk 9]
```diff
@@ -24,18 +24,17 @@
...
-    def set_address(self, state, address, container):
-        try:
-            return state.set_state(
-                entries={address: container.SerializeToString()},
-                timeout=TIMEOUT_SECONDS,
-            )
-        except FutureTimeoutError:
-            raise InternalError(ERROR_MESSAGE_TIMEOUT, TIMEOUT_SECONDS)
+def get_address(state, address):
+    try:
```
Comment (nadiabahrami) [inline rbac/common/sawtooth/state_client.py]: docstrings

File: rbac/common/sawtooth/state_client.py [hunk 10]
```diff
@@ -24,18 +24,17 @@
...
+        return state.get_state(addresses=[address], timeout=TIMEOUT_SECONDS)
+    except FutureTimeoutError:
+        raise InternalError(ERROR_MESSAGE_TIMEOUT, TIMEOUT_SECONDS)
+
+
+def set_address(state, address, container):
+    try:
```
Comment (nadiabahrami) [inline rbac/common/sawtooth/state_client.py]: DS

File: rbac/processor/event_handler.py [hunk 11]
```diff
@@ -119,8 +125,19 @@ def namespaces(self):
         return addresser.family.namespaces
 
     def apply(self, transaction, state):
-        payload = RBACPayload()
-        payload.ParseFromString(transaction.payload)
+        try:
```
Comment (nadiabahrami) [inline rbac/processor/event_handler.py]: Docstrings on this whole file

Comment (adamgering) [reply rbac/processor/event_handler.py]: Isn't my code, this is legacy code that will soon disappear.

Comment (nadiabahrami) [reply rbac/processor/event_handler.py]: cool if it is legacy this one is fine to not do docstrings

File: rbac/common/sawtooth/batcher.py (hunk 8 above)
Comment (adamgering) [reply rbac/common/sawtooth/batcher.py]: It's not worth the time to rebase #536 to add docstrings to #532.
Docstrings will be added in a followup PR.

File: rbac/common/sawtooth/state_client.py (hunk 9 above)
Comment (adamgering) [reply rbac/common/sawtooth/state_client.py]: It's not worth the time to rebase #536 to add docstrings to #532.
Docstrings will be added in a followup PR.

File: rbac/common/sawtooth/state_client.py (hunk 10 above)
Comment (adamgering) [reply rbac/common/sawtooth/state_client.py]: It's not worth the time to rebase #536 to add docstrings to #532.
Docstrings will be added in a followup PR.

File: rbac/common/sawtooth/__init__.py (hunk 2 above)
Comment (kthblmfld) [reply rbac/common/sawtooth/__init__.py]: It specifies what will be pulled in when a 'from sawtooth import *' is performed, right?  So it filters out any additional variables that would otherwise be pulled in on a wildcard import.

File: rbac/common/base/base_message.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kthblmfld) [reply rbac/common/base/base_message.py]: Understood. This is an improvement over that. I am just not sure about having a base class register with another module and then be extended and ultimately instantiated. It confuses me because I understand a base class as nothing more than common features to be customized as needed in subclassing for operations on the interface (polymorphism).

I could be missing something here. We should discuss.

File: rbac/common/base/base_message.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kthblmfld) [reply rbac/common/base/base_message.py]: Very well. The offending change was already merged. 

We need to find some middle ground between the need to correct offending code retroactively and the blocking of PRs that break code that does not exist in the repo yet.

File: rbac/common/sawtooth/batcher.py (hunk 8 above)
Comment (kthblmfld) [reply rbac/common/sawtooth/batcher.py]: @adamgering, you have blocked PRs in the past week over:
  - whitespace change to a file, performed automatically by Black  
  - removal of the 3-4th layer of env variable redundancy having no effect on the master branch

Those were apparently worth the time of others.

We have agreed to docstring our code.

File: rbac/common/base/base_message.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/common/base/base_message.py]: The base class isn't registering, the parent classes are.

File: rbac/common/sawtooth/__init__.py (hunk 2 above)
Comment (adamgering) [reply rbac/common/sawtooth/__init__.py]: It marked the variable for purposes of export. It may not be required here, but it is good practice. It is required when you import a variable that you intend to export, as the linter will suggest it is an unused variable, and the app will break when you remove it.

File: rbac/common/sawtooth/batcher.py (hunk 8 above)
Comment (adamgering) [reply rbac/common/sawtooth/batcher.py]: I asked for PRs not to change files that have nothing to do with the PR.
I asked for a PR not to remove a function that was added for a required purpose.
Blocking PRs as a retaliation seems pretty and unprofessional and puts the deliverables at risk. Docstring are not yet required by our lint standards.
